### ISSUE TITLE
Allow regexes in RuleTester

### ISSUE THREAD

**Issue by Daniel15** [eslint/eslint#7837]
* **ESLint Version:** 3.5.0

It would be nice for the RuleTester to allow regular expressions to match messages, rather than just strings. At Facebook, some of our lint rules have very long messages - Here's one such message from a lint rule that has several possible messages:
```js
context.report({
  node: node,
  message:
    'This is a shared module but it is referencing a ' +
    '{{dependencyProduct}} module ({{requiredModule}}). Shared ' +
    'modules should not have any dependencies on product code. ' +
    'If this module is specific to {{dependencyProduct}}, it should ' +
    'be moved to {{productPath}}. Contact danlo@ if this is a ' +
    'false positive.',
  data: {
    dependencyProduct: dependencyProduct,
    productPath: getPathForProductName(dependencyProduct),
    requiredModule: requiredModule,
  },
});
```

And an example of one of the test cases:
```js
{
  // Shared module requiring a Power Editor module
  code: 'require("AdsPEApplication")',
  filename: 'html/js/ads/utils/AdsBidUtils.js',
  errors: [
    'This is a shared module but it is referencing a Power Editor module ' +
    '(AdsPEApplication). Shared modules should not have any dependencies ' +
    'on product code. If this module is specific to Power Editor, it ' +
    'should be moved to html/js/ads/powereditor. Contact danlo@ if this ' +
    'is a false positive.',
  ],
},
```

All I really want to check is that the correct message is being returned, and the correct variables are merged into it. Right now I need to copy and paste the entire message for every test case. If the message changes at all, I need to remember to update every test case. Alternatively, I could export the message from the lint rule module and import it into the unit test, but that feels a bit messy and I'd still need to deal with the variables in the string.

It would be nicer to be able to use a regex, so I could do something like:
```
  errors: [
    /Power Editor module \(AdsPEApplication\).+should be moved to html\/js\/ads\/powereditor/,
  ],
```

Alternatively, it would also be useful to be able to provide a "message ID" to the `context.report` call:
```js
context.report({
  node: node,
  message: 'Hello world',
  message_id: 2, // Any arbitrary integer
});
```

Such that the RuleTester could match the message by ID rather than by actual message text:
```js
{
  code: '// Foo bar',
  errors: [
    {id: 2}, // "2" matches the integer used in context.report
  ],
}
```

**Comment by vitorbal** [comment]
Hi @Daniel15, thanks for the suggestion! I'm not sure about the message ID part, but the regex sounds like a nice improvement. Let's see what some of the other team members think.

Would you be willing to send a PR for this if it gets accepted?

**Comment by ilyavolodin** [comment]
@Daniel15 If your goal is to just verify that correct variables are being inserted into the message, you could just check it once in your tests, and skip `errors` property for all of the other tests (or use `errors: 1` instead. Errors property can accept numbers as well as text). I'm not really opposed to adding support for RegExp, but we are trying to keep RuleTester as simple as possible, to avoid additional maintenance.

**Comment by nzakas** [comment]
I'm 👍 for supporting regexes anywhere we currently support strings for checking error messages, :-1: for the message ID thing. 

@Daniel15 if accepted, are you willing to submit a PR for this?

**Comment by Daniel15** [comment]
> if accepted, are you willing to submit a PR for this?

Sure, I can do that!

Another alternative I thought of is to provide a function where you can add your own checker logic, for example something like:
```
{
  // Shared module requiring a Power Editor module
  code: 'require("AdsPEApplication")',
  filename: 'html/js/ads/utils/AdsBidUtils.js',
  errors: [
    error => /Power Editor module \(AdsPEApplication\)/.test(error.message),
  ],
},
```

> If your goal is to just verify that correct variables are being inserted into the message, you could just check it once in your tests, and skip errors property for all of the other tests (or use errors: 1 instead. 

@ilyavolodin - There's multiple different error messages that can be thrown - Three unique messages in the case of this particular lint rule. For every test case, I want to verify that the correct message (out of the three possibilities) is thrown, and that all the correct variables are merged in.

**Comment by nzakas** [comment]
Let's stick with regexes for this first go around. We can always extend it to take functions later if we think that's necessary, id just like to keep this as small as possible to start.

**Comment by nzakas** [comment]
TSC Summary: proposal is to allow regular expressions in RuleTester for testing that error messages are correct. Anywhere a string is currently supported for this purpose should also accept a regex. 

TSC Question: do we want to accept this enhancement?

**Comment by alberto** [comment]
This has been accepted by TSC

**Comment by nzakas** [comment]
@Daniel15 you're all clear to begin a PR for regex message checks.

**Comment by platinumazure** [comment]
@Daniel15 Friendly ping. Are you still interested in writing a PR for this?

**Comment by Daniel15** [comment]
Hi! Sorry for the delay, I got busy with other commitments. I'll try to
take a look this weekend :)

Sent from my phone.

On Feb 6, 2017 7:00 PM, "Kevin Partington" <notifications@github.com> wrote:

@Daniel15 <https://github.com/Daniel15> Friendly ping. Are you still
interested in writing a PR for this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/eslint/eslint/issues/7837#issuecomment-277887329>, or mute
the thread

**Comment by Daniel15** [comment]
I haven't forgotten about this, I'm still going to submit a PR soon.

**Comment by Daniel15** [comment]
#8115
